### ISSUE TITLE
Segfault in 0.16.0 - Unref assertion error assert(refs_ > 0)

### ISSUE THREAD

**Issue by rc0x03** [polotek/libxmljs#369]
I'm getting a segfault in the latest version. I believe it's related to using "addChild" between documents, however I have to do further investigating. It's being caused by obfuscated JS code from a web page, so it's a real pain to track down. The cause is either a double Unref or a call to Unref but not Ref. I'll update this issue once I create a test for it.

**Comment by teleological** [comment]
Have you been able to determine whether the failing `Unref()` is for a document or a node?

**Comment by rc0x03** [comment]
I don't have access to my server right now, but from my brief investigation this morning I'm pretty sure it's Unrefing an ancestor node. After disabling the addChild call, the code could continue without the segfault (but not very far since it depends on the addChild call to run). If I recall correctly, the code was appending a `<table>` with two `<td>` elements to a `<div>`. When the garbage collector would run the destructor for `<td>` the `xml_obj` would already be freed and would segfault when it tried to Unref the stored wrapped ancestor.

**Comment by teleological** [comment]
I think I see the problem. `ancestor` is a pointer to an `xmlNode` (i.e. not a wrapper) which may already be freed, yet we assume we can access its `_private` member.

**Comment by rc0x03** [comment]
Although in this case it seems the node being freed is able to access `ancestor` and `ancestor->_private` without any issues. The problem occurs only when it tries to Unref the ancestor. ObjectWrap throws an assertion error because we tried to Unref but ref_ is already 0. That being said, it is strange that `ancestor` isn't null but the child `xml_obj` is.

**Comment by teleological** [comment]
Helpful clarification. Please share a test case if you develop one.

**Comment by teleological** [comment]
Case for null `xml_obj`, non-null `ancestor`: Graph `A-B-C-D` such that `B` and `D` are wrapped, and `D` ref'd ancestor `B`. `C` is then detached. If `C` and `D` are unreferenced, and `C` is destroyed first, then when the destructor runs for `D`, `xml_obj` will be NULL because of deregistration, but `ancestor` will not be, and needs to be unref'd to prevent `B` from leaking.

**Comment by teleological** [comment]
Any progress? Or sense of how easy/hard it is to replicate this condition?

**Comment by rc0x03** [comment]
I haven't looked back on the original problem yet. I was curious to see if I could recreate it with a fuzzer instead of looking in to the obfuscated code. I was able to find a few more weird segfaults and hang ups while fuzzing the libxml. I'm going to upload `libxml-tree-fuzzer` And add you as a contributor.

**Comment by rc0x03** [comment]
The issue seems to be caused by setting a node's text even though it has children.

```javascript
var libxmljs = require('libxmljs');

var segfault_test = 1;

var child = (function() {
    var doc = libxmljs.parseHtml('<root></root>')
    var root = doc.root();
    var child = new libxmljs.Element(doc, 'child').text('content');

    if (segfault_test == 1) {
        child.attr({ test: true })
        var attr = child.attr('test'); // this will cause a segfault during GC
    }

    if (segfault_test == 2) {
        var inner = new libxmljs.Element(doc, 'inner');
        child.addChild(inner); // this will also cause a segfault during GC
    }

    root.addChild(child);
    root.text('');

    return child;
})()

global.gc();

child.toString(); // returns "", but should still return "<child>content</child>"

if (segfault_test == 3)
    child.parent(); // this will also cause a segfault

if (segfault_test == 4)
    child.childNodes(); // another way to segfault
```

**Comment by rc0x03** [comment]
The workaround is to manually loop through each child and remove() it before calling `Element.text(..)`. I think that libxmljs should automatically remove child nodes when setting the inner text.

**Comment by rc0x03** [comment]
@teleological I've added you as a collaborator on `libxmljs-fuzzer`

**Comment by teleological** [comment]
@rc0x03 implemented your suggestion to automatically remove displaced children in #373. fuzzer looks like it could be useful for shaking other bugs out.

**Comment by rc0x03** [comment]
Fixed with #373
